### ISSUE TITLE
[REQEST] Add support for SMARTMI Silent Smart Air Humidifier 2 (With Display) "zhimi.humidifier.ca4"

### ISSUE THREAD

**Issue by DJay-X** [russtone/homebridge-mi-humidifier#26]
I hope to get support with this plugin for the SMARTMI Smart Air Humidifier 2 (With Display) "zhimi.humidifier.ca4"with Display. Youpin Smartmi CJXJSQ04ZM
Thanks in advance to all those who contribute to its implementation.
If I as a user can do something to help, please of course let me know.

![unnamed](https://user-images.githubusercontent.com/3874239/79687427-72e32880-8247-11ea-82f5-c91219483f83.png)

**Comment by DJay-X** [comment]
@russtone Today I received my new SMARTMI Smart Air Humidifier 2 (With Display) 
zhimi-humidifier-ca4-mibtC33D

As expected the plugin does not support it at the moment.
If there is anything I can do to help to get it to work, please of course let me know.

Thanks again for your work, the time you are investing and sharing the plugin with us.

**Comment by Dane03** [comment]
Possible to ‚fake‘ it being the old Model?
No features that would be lost...

Maybe that works out of the box?

**Comment by Dane03** [comment]
@russtone any news for this model?

**Comment by russtone** [comment]
Hey @Dane03,
I can't add support for this model now, because as I know it uses miotspec protocol, which has slightly different commands and I just don't know this commands. Maybe if you can collect some network traffic dumps It can help.

**Comment by ileodo** [comment]
@russtone looks like this plugin works for ca4 model.
https://www.npmjs.com/package/homebridge-smartmi-humidifier2 
I'll have a look at the source code and try to contribute something in your project. It would be great if you have have a look as well.

**Comment by ileodo** [comment]
@russtone found the spec for this model: [http://miot-spec.org/miot-spec-v2/instance?type=urn:miot-spec-v2:device:humidifier:0000A00E:zhimi-ca4:2](http://miot-spec.org/miot-spec-v2/instance?type=urn:miot-spec-v2:device:humidifier:0000A00E:zhimi-ca4:2)

**Comment by russtone** [comment]
@ileodo Thanks for the information! I will have a look at it

**Comment by ileodo** [comment]
@russtone I have given a try yesterday, but couldn't fit it in the current code structure. Are you open for some code restructure?

**Comment by ileodo** [comment]
Please have a look at my PR if possible.

**Comment by DJay-X** [comment]
@russtone 
@ileodo 
Thanks to you guys for your work.
Unfortunately it's not working for me with my ca4

```
        {
            "name": "Luftbefeuchter",
            "ip": "192.168.178.xx",
            "token": "00000000000000000000000000000000",
            "model": "ca4",
            "showMuteSwitch": false,
            "nameMuteSwitch": "Mute humidifier",
            "showTemperature": true,
            "nameTemperature": "Luftbefeuchter Temperatur",
            "showHumidity": true,
            "nameHumidity": "Luftbefeuchter",
            "accessory": "MiHumidifier"
        }
```

```
Jul 27 13:13:10 raspberrypi-3BPlus homebridge[24619]: [2020-7-27 13:13:10] Loaded plugin: homebridge-mi-humidifier@1.3.0
Jul 27 13:13:10 raspberrypi-3BPlus homebridge[24619]: [2020-7-27 13:13:10] Registering accessory 'homebridge-mi-humidifier.MiHumidifier'
Jul 27 13:13:23 raspberrypi-3BPlus homebridge[24619]: [2020-7-27 13:13:23] TypeError: Cannot read property 'powerGetName' of undefined
Jul 27 13:13:23 raspberrypi-3BPlus homebridge[24619]:     at new MiHumidifier (/usr/local/lib/node_modules/homebridge-mi-humidifier/index.js:55:129)
Jul 27 13:13:23 raspberrypi-3BPlus homebridge[24619]:     at /usr/local/lib/node_modules/homebridge/lib/server.js:246:39
Jul 27 13:13:23 raspberrypi-3BPlus homebridge[24619]:     at Array.forEach (<anonymous>:null:null)
Jul 27 13:13:23 raspberrypi-3BPlus homebridge[24619]:     at Server._loadAccessories (/usr/local/lib/node_modules/homebridge/lib/server.js:222:33)
Jul 27 13:13:23 raspberrypi-3BPlus homebridge[24619]:     at Server.start (/usr/local/lib/node_modules/homebridge/lib/server.js:76:18)
Jul 27 13:13:23 raspberrypi-3BPlus homebridge[24619]:     at cli (/usr/local/lib/node_modules/homebridge/lib/cli.js:88:12)
Jul 27 13:13:23 raspberrypi-3BPlus homebridge[24619]:     at Object.<anonymous> (/usr/local/lib/node_modules/homebridge/bin/homebridge:17:22)
Jul 27 13:13:23 raspberrypi-3BPlus homebridge[24619]:     at Module._compile (internal/modules/cjs/loader.js:1137:30)
Jul 27 13:13:23 raspberrypi-3BPlus homebridge[24619]:     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1157:10)
Jul 27 13:13:23 raspberrypi-3BPlus homebridge[24619]:     at Module.load (internal/modules/cjs/loader.js:985:32)
Jul 27 13:13:23 raspberrypi-3BPlus homebridge[24619]:     at Function.Module._load (internal/modules/cjs/loader.js:878:14)
Jul 27 13:13:23 raspberrypi-3BPlus homebridge[24619]:     at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:71:12)
Jul 27 13:13:23 raspberrypi-3BPlus homebridge[24619]:     at internal/main/run_main_module.js:17:47
Jul 27 13:13:24 raspberrypi-3BPlus homebridge[24619]: [2020-7-27 13:13:24] Got SIGTERM, shutting down Homebridge...
```

**Comment by russtone** [comment]
@DJay-X I've updated npm version, please try now

**Comment by ileodo** [comment]
Moving forward, if you think so, we can have all new model data driven. And if anybody want,we can start to migrate the existing model as well.

**Comment by DJay-X** [comment]
Up and running. 👍  Thanks a lot. 😃

**Comment by DJay-X** [comment]
@russtone @IldarMurzenkov 
Hi Guys. I hope you are all doing well.
I recently noticed that I see in the log after a homebridge restart the following message.
The humidifier is on and connected to Wifi of course. 
And after the finish of reboot of homebridge, I also can control hit.
However, I'm worried about the "Fail to discover the device." message at start.
Any Idea?
Thanks in advance.

```
Okt 02 09:12:50 raspberrypi-3BPlus homebridge[12686]: [2020-10-2 9:12:50] [Luftbefeuchter] Fail to discover the device. Retry in 1 minute Error: Could not connect to device, handshake timeout
Okt 02 09:12:50 raspberrypi-3BPlus homebridge[12686]:     at Timeout._onTimeout (/usr/local/lib/node_modules/homebridge-mi-humidifier/node_modules/miio/lib/network.js:427:17)
Okt 02 09:12:50 raspberrypi-3BPlus homebridge[12686]:     at listOnTimeout (internal/timers.js:549:17)
Okt 02 09:12:50 raspberrypi-3BPlus homebridge[12686]:     at processTimers (internal/timers.js:492:7)
Okt 02 09:12:50 raspberrypi-3BPlus homebridge[12686]:  {
Okt 02 09:12:50 raspberrypi-3BPlus homebridge[12686]:   code: 'timeout',
Okt 02 09:12:50 raspberrypi-3BPlus homebridge[12686]:   device: null
Okt 02 09:12:50 raspberrypi-3BPlus homebridge[12686]: }
```

**Comment by DJay-X** [comment]
@russtone I still see the above error/warning message every time I restart homebridge. 
`Fail to discover the device. Retry in 1 minute Error:`
Is there anything I can do from my side? The Humidifier is of course always on power and connected to wifi.
Thanks in advance.

**Comment by russtone** [comment]
hey @DJay-X,
When the plugin is started it sends UDP packet with handshake message to the device and waits for response.
`Fail to discover the device` error occurs if device is not responding or responds for a long time. 
I don't know why this happens, but you can collect

**Comment by DJay-X** [comment]
> Do you run homebridge using systemd?
Yes, I do run homebridge using systemd.

**Comment by russtone** [comment]
And you get error message after restarting service using `systemctl restart`?

**Comment by DJay-X** [comment]
Yesterday I switched from systemd to hb-service.
Same message in the log when I restart homebridge.

```
[21.11.2020, 13:15:58] [HB Supervisor] Homebridge Process Ended. Code: 143, Signal: null
[21.11.2020, 13:16:03] [HB Supervisor] Restarting Homebridge...
[21.11.2020, 13:16:03] [HB Supervisor] Starting Homebridge with extra flags: -I
[21.11.2020, 13:16:03] [HB Supervisor] Started Homebridge v1.1.6 with PID: 12097
[21.11.2020, 13:16:04] Loaded config.json with 23 accessories and 17 platforms.
[21.11.2020, 13:16:16] Loaded plugin: homebridge-mi-humidifier@1.4.0
[21.11.2020, 13:16:16] Registering accessory 'homebridge-mi-humidifier.MiHumidifier'
[21.11.2020, 13:16:25] [Luftbefeuchter] Fail to discover the device. Retry in 1 minute Error: Could not connect to device, handshake timeout
    at Timeout._onTimeout (/usr/local/lib/node_modules/homebridge-mi-humidifier/node_modules/miio/lib/network.js:427:17)
    at listOnTimeout (internal/timers.js:554:17)
    at processTimers (internal/timers.js:497:7)
 {
  code: 'timeout',
  device: null
}
```

However, also then before the Humidifier is responding. Strange the Timeout error is showing in the log on start.

**Comment by koles0ff** [comment]
Hello guys,

would vote for this feature as owner of the v. 2 of the device. Currently using other plugin which is honestly sucks and mostly useless (Homebridge Smartmi Humidifier2)

**Comment by DJay-X** [comment]
@koles0ff Hey. This plugin here works already well with the v2 with Display. 
I only still get the "[Fail to discover the device](https://github.com/russtone/homebridge-mi-humidifier/issues/26#issuecomment-731571671)." every time making a homebridge restart. However controlling the humidifier works.

**Comment by DJay-X** [comment]
> [21.11.2020, 13:16:25] [Luftbefeuchter] Fail to discover the device. Retry in 1 minute Error: Could not connect to device, handshake timeout
>     at Timeout._onTimeout (/usr/local/lib/node_modules/homebridge-mi-humidifier/node_modules/miio/lib/network.js:427:17)
>     at listOnTimeout (internal/timers.js:554:17)
>     at processTimers (internal/timers.js:497:7)
>  {
>   code: 'timeout',
>   device: null
> }

This seems to be fixed with v2.0.0-beta and switching to "platforms" 👍
